File: grafos/GrafoPonderado.cpp
#include <iostream>
#include <iomanip>

const int V = 4;
int matriz[V][V];

void agregar_arista(int a, int b, int peso) {
    // Agrega una arista ponderada no dirigida entre a y b
    matriz[a][b] = peso;
    matriz[b][a] = peso;
}

void imprimir_matriz_con_bordes() {
    // Imprime la matriz de adyacencia con bordes (como tabla)
    std::cout << "+";
    for (int i = 0; i < V + 1; ++i) {
        std::cout << "----+";
    }
    std::cout << std::endl;

    std::cout << "|    |";
    for (int i = 0; i < V; ++i) {
        std::cout << " " << i << " |";
    }
    std::cout << std::endl;

    std::cout << "+";
    for (int i = 0; i < V + 1; ++i) {
        std::cout << "----+";
    }
    std::cout << std::endl;

    for (int i = 0; i < V; ++i) {
        std::cout << "| " << i << "  |";
        for (int j = 0; j < V; ++j) {
            std::cout << " " << std::setw(2) << matriz[i][j] << " |";
        }
        std::cout << std::endl;

        std::cout << "+";
        for (int k = 0; k < V + 1; ++k) {
            std::cout << "----+";
        }
        std::cout << std::endl;
    }
}

void bfs(int inicio) {
    // BFS adaptado para grafo ponderado (ignora pesos, solo conectividad)
    bool visitado[V] = {false};
    int cola[V];
    int frente = 0;
    int fondo = 0;
    cola[fondo++] = inicio;
    visitado[inicio] = true;

    while (frente < fondo) {
        int actual = cola[frente++];
        std::cout << actual << " ";

        for (int vecino = 0; vecino < V; ++vecino) {
            if (matriz[actual][vecino] > 0 && !visitado[vecino]) {
                cola[fondo++] = vecino;
                visitado[vecino] = true;
            }
        }
    }
}

void dfs(int inicio) {
    // DFS adaptado para grafo ponderado (ignora pesos)
    bool visitado[V] = {false};
    dfs_recursivo(inicio, visitado);
    std::cout << std::endl;
}

void dfs_recursivo(int actual, bool visitado[]) {
    visitado[actual] = true;
    std::cout << actual << " ";
    for (int vecino = 0; vecino < V; ++vecino) {
        if (matriz[actual][vecino] > 0 && !visitado[vecino]) {
            dfs_recursivo(vecino, visitado);
        }
    }
}

int main() {
    // Inicializar matriz
    for (int i = 0; i < V; ++i) {
        for (int j = 0; j < V; ++j) {
            matriz[i][j] = 0;
        }
    }

    // Agregar aristas ponderadas
    agregar_arista(0, 1, 4);  // 0 —4— 1
    agregar_arista(0, 2, 2);  // 0 —2— 2
    agregar_arista(1, 3, 1);  // 1 —1— 3
    agregar_arista(2, 3, 3);  // 2 —3— 3

    // Mostrar resultados
    std::cout << "Matriz de Adyacencia (representación del grafo ponderado):\n";
    imprimir_matriz_con_bordes();
    std::cout << "BFS desde 0: ";
    bfs(0);
    std::cout << "DFS desde 0: ";
    dfs(0);

    return 0;
}

File: grafos/GrafoPonderado.py
# Grafo Ponderado
# Un grafo ponderado es un grafo donde cada arista tiene un "peso" (número), representando costo, distancia o tiempo.
# Por qué es así: Permite modelar escenarios reales con valores cuantitativos, como distancias en mapas (km) o costos en redes (dinero).
# Requisitos:
# - Matriz usa enteros >0 para pesos (en vez de 1/0).
# - 0 significa no arista.
# - BFS/DFS se adaptan (ignoran pesos, solo conectividad).
# - En código: agregar_arista toma peso como parámetro.

V = 4
matriz = [[0 for _ in range(V)] for _ in range(V)]

def agregar_arista(a, b, peso):
    """Agrega una arista ponderada no dirigida entre a y b."""
    matriz[a][b] = peso
    matriz[b][a] = peso

def imprimir_matriz_con_bordes():
    """Imprime la matriz de adyacencia con bordes (como tabla)."""
    print("+", end="")
    for _ in range(V + 1):
        print("----+ ", end="")
    print()

    print("|    |", end="")
    for i in range(V):
        print(f" {i} |", end="")
    print()

    print("+", end="")
    for _ in range(V + 1):
        print("----+ ", end="")
    print()

    for i in range(V):
        print(f"| {i}  |", end="")
        for j in range(V):
            print(f" {matriz[i][j]:>2} |", end="")
        print()

        print("+", end="")
        for _ in range(V + 1):
            print("----+ ", end="")
        print()

def bfs(inicio):
    """BFS adaptado para grafo ponderado (ignora pesos, solo conectividad)."""
    visitado = [False] * V
    cola = [0] * V
    frente = 0
    fondo = 0
    cola[fondo] = inicio
    fondo += 1
    visitado[inicio] = True

    while frente < fondo:
        actual = cola[frente]
        frente += 1
        print(actual, end=" ")

        for vecino in range(V):
            if matriz[actual][vecino] > 0 and not visitado[vecino]:
                cola[fondo] = vecino
                fondo += 1
                visitado[vecino] = True

def dfs(inicio):
    """DFS adaptado para grafo ponderado (ignora pesos)."""
    visitado = [False] * V
    dfs_recursivo(inicio, visitado)
    print()

def dfs_recursivo(actual, visitado):
    visitado[actual] = True
    print(actual, end=" ")
    for vecino in range(V):
        if matriz[actual][vecino] > 0 and not visitado[vecino]:
            dfs_recursivo(vecino, visitado)

# --- Ejecución principal ---
if __name__ == "__main__":
    # Inicializar matriz
    for i in range(V):
        for j in range(V):
            matriz[i][j] = 0

    # Agregar aristas ponderadas
    agregar_arista(0, 1, 4)  # 0 —4— 1
    agregar_arista(0, 2, 2)  # 0 —2— 2
    agregar_arista(1, 3, 1)  # 1 —1— 3
    agregar_arista(2, 3, 3)  # 2 —3— 3

    # Mostrar resultados
    print("Matriz de Adyacencia (representación del grafo ponderado):")
    imprimir_matriz_con_bordes()
    print("BFS desde 0: ", end="")
    bfs(0)
    print("DFS desde 0: ", end="")
    dfs(0)